core/utils/memory_utils.h

```diff
@@ -69,6 +70,37 @@ namespace memoryUtils
     inline constexpr T GB(T bytes) {
       return  bytes / ((T)1024 * (T)1024 * (T)1024);
     }
+
+    /** Returns the current alignment of the ptr.
+     * Essentially just counts the amount of LSB zeros.
+     * Returns 2^31 for ptr == 0. */
+	inline uint32_t alignmentOf(uintptr_t ptr) {
+		for (uint32_t alignPower = 0; alignPower < 32; alignPower++) {
+			const uint32_t alignment = 1u << alignPower;
+			if ((ptr & alignment) != 0) {
+				return alignment;
+			}
+		}
+		return 1u << 31u;
+	}
+
+	/** Convenience shortcut for alignmentOf(uintptr_t). */
+	inline uint32_t alignmentOf(void * ptr) {
+		return alignmentOf((size_t) ptr);
+	}
+
+	/** Return number + N, where N is the smallest non-negative number required to make number have given alignment.
+	 * Example: align(13, 8) = 16 */
+	template<typename T>
```

Comment:
I don't understand what this function does.

Reply:
Imagine writing a custom [linear allocator](https://github.com/mtrebi/memory-allocators#linear-allocator) which supports alignment:
```c++
void* alloc(size_t size, uint32_t alignment) {
    void* start = align(linear_alloc.offset, alignment);
    linear_alloc.offset = start + size;
    return start;
}
```

I am using this to allocate more memory than strictly necessary to allow vectorized operations to not worry about operating on invalid memory for the last group of operands. E.g. in case where I have an array of `int`s and I want to operate on them in groups of 4, I can do something like: (CUDA-like pseudocode)
```c++
void do_stuff(int amount) {
    int* data = new int[align(amount, 4)];
    fill(data, amount);
    int4* data4 = (int4*)data;
    for (int i = 0; i < align(amount, 4); i++) {
        // It is easier and faster to do up to 3 extra additions than to add extra loop for residue
        data4[i] += 3;
    }
}
```